src/js/tracing/nativeframes.ts

```diff
@@ -0,0 +1,263 @@
+import { Span, Transaction } from "@sentry/tracing";
+import { Event, EventProcessor, Measurements } from "@sentry/types";
+import { logger, timestampInSeconds } from "@sentry/utils";
+
+import { NativeFramesResponse } from "../definitions";
+import { NATIVE } from "../wrapper";
+import { instrumentChildSpanFinish } from "./utils";
+
+interface FramesMeasurements extends Measurements {
+  frames_total: {
+    value: number;
+  };
+  frames_slow: {
+    value: number;
+  };
+  frames_frozen: {
+    value: number;
+  };
+}
+
+const MARGIN_OF_ERROR_SECONDS = 0.05;
+
+/**
+ * Instrumentation to add native slow/frozen frames measurements onto transactions.
+ */
+export class NativeFramesInstrumentation {
+  private _finishFrames: Map<
+    string,
+    { timestamp: number; nativeFrames: NativeFramesResponse | null }
+  > = new Map();
+  private _framesListeners: Map<string, () => void> = new Map();
+  private _lastSpanFinishFrames?: {
+    timestamp: number;
+    nativeFrames: NativeFramesResponse;
+  };
+
+  public constructor(
+    addGlobalEventProcessor: (e: EventProcessor) => void,
+    doesExist: () => boolean
+  ) {
+    logger.log(
+      "[ReactNativeTracing] Native frames instrumentation initialized."
+    );
+
+    addGlobalEventProcessor((event) => {
+      if (doesExist()) {
+        return this._processEvent(event);
+      }
+
+      return event;
+    });
+  }
+
+  /**
+   * To be called when a transaction is started
+   */
+  public onTransactionStart(transaction: Transaction): void {
+    void NATIVE.fetchNativeFrames().then((framesMetrics) => {
+      if (framesMetrics) {
+        transaction.setData("__startFrames", framesMetrics);
```

Comment:
Why the `__` prefix?

Reply:
`__startFrames` is just to be used by this instrumentation and removed on the transaction finish. It's to make it clear that this is internal I guess.

Reply:
Ah hmmm, thought that this was kept and not removed on finish. Feels weird to use the event as a carrier for temporary data. Is there no way around this?

@bruno-garcia what do you think?

Reply:
Doing the `transaction.data.__startFrames` getter feels like a code smell to me.

Reply:
We can just store it in a map then.

Reply:
@AbhiPrasad Ok I've given it a go and using a map means we could have a leak, the transaction could be started but never finished or lost and we would still store the start frames. A weakmap can't be used due to us not having a reference to the transaction inside the event processor that needs to access the start frames. I guess using the event as a carrier as is would work best as it covers this edge case?

Reply:
Ah hmm. I mean it's not the end of the world to keep it on the event, especially if the alternative has possible problems. Let's just use the transaction data then.